src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActions/SuggestedAction.cs

```diff
@@ -96,80 +96,76 @@ protected Task<ImmutableArray<CodeActionOperation>> GetPreviewOperationsAsync(Ca
+            // we're going to return immediately from Invoke and kick off our own async work to invoke the
+            // code action. Once this returns, the editor will close the threaded wait dialog it created.
+            // So we need to take ownership of it and start our own TWD instead to track this.
+            context.TakeOwnership();
+
+            _ = InvokeAsync();
         }
 
-        private void Invoke(IProgressTracker progressTracker, CancellationToken cancellationToken)
+        private async Task InvokeAsync()
         {
-            // While we're not technically doing anything async here, we need to let the
-            // integration test harness know that it should not proceed until all this
-            // work is done.  Otherwise it might ask to do some work before we finish.
-            // That can happen because although we're on the UI thread, we may do things
-            // that end up causing VS to pump the messages that the test harness enqueues
-            // to the UI thread as well.
-            using (SourceProvider.OperationListener.BeginAsyncOperation($"{nameof(SuggestedAction)}.{nameof(Invoke)}"))
+            try
+            {
+                using var token = SourceProvider.OperationListener.BeginAsyncOperation($"{nameof(SuggestedAction)}.{nameof(Invoke)}");
+                using var context = SourceProvider.UIThreadOperationExecutor.BeginExecute(
+                    EditorFeaturesResources.Execute_Suggested_Action, CodeAction.Title, allowCancellation: true, showProgress: true);
+                using var scope = context.AddScope(allowCancellation: true, CodeAction.Message);
+                await this.InnerInvokeAsync(new UIThreadOperationContextProgressTracker(scope), context.UserCancellationToken).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+            }
+            catch (Exception ex) when (FatalError.ReportAndCatch(ex))
             {
-                InnerInvoke(progressTracker, cancellationToken);
             }
         }
 
-        protected virtual void InnerInvoke(IProgressTracker progressTracker, CancellationToken cancellationToken)
+        protected virtual async Task InnerInvokeAsync(IProgressTracker progressTracker, CancellationToken cancellationToken)
         {
-            AssertIsForeground();
+            await this.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
 
-            var snapshot = SubjectBuffer.CurrentSnapshot;
             using (new CaretPositionRestorer(SubjectBuffer, EditHandler.AssociatedViewService))
             {
-                Document getFromDocument() => SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
-                InvokeCore(getFromDocument, progressTracker, cancellationToken);
+                // ConfigureAwait(true) so that CaretPositionRestorer.Dispose runs on the UI thread.
```

Comment:
i know this concerned you.  however, this is the only guy that has a need for CA(true) so that the Dispose runs on the UI thread.